File: apps/mobile/lib/features/social/data/social_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

class SocialConnection {
  final String id;
  final String friendId;
  final String friendName;
  final String friendEmail;
  final String? friendAvatar;
  final String status; // pending | accepted | blocked
  final String direction; // outgoing | incoming
  final String createdAt;

  const SocialConnection({
    required this.id,
    required this.friendId,
    required this.friendName,
    required this.friendEmail,
    this.friendAvatar,
    required this.status,
    required this.direction,
    required this.createdAt,
  });

  factory SocialConnection.fromMap(Map<String, dynamic> m) => SocialConnection(
        id: m['id'] as String? ?? '',
        friendId: m['friend_id'] as String? ?? '',
        friendName: m['friend_name'] as String? ?? 'Unknown',
        friendEmail: m['friend_email'] as String? ?? '',
        friendAvatar: m['friend_avatar'] as String?,
        status: m['status'] as String? ?? 'pending',
        direction: m['direction'] as String? ?? 'outgoing',
        createdAt: m['created_at'] as String? ?? '',
      );
}

class FriendActivity {
  final String friendId;
  final String friendName;
  final String? friendAvatar;
  final int habitsToday;
  final int tasksToday;
  final int? momentumScore;
  final String? streakHighlight;

  const FriendActivity({
    required this.friendId,
    required this.friendName,
    this.friendAvatar,
    required this.habitsToday,
    required this.tasksToday,
    this.momentumScore,
    this.streakHighlight,
  });

  factory FriendActivity.fromMap(Map<String, dynamic> m) => FriendActivity(
        friendId: m['friend_id'] as String? ?? '',
        friendName: m['friend_name'] as String? ?? 'Unknown',
        friendAvatar: m['friend_avatar'] as String?,
        habitsToday: (m['habits_today'] as num?)?.toInt() ?? 0,
        tasksToday: (m['tasks_today'] as num?)?.toInt() ?? 0,
        momentumScore: (m['momentum_score'] as num?)?.toInt(),
        streakHighlight: m['streak_highlight'] as String?,
      );
}

class InviteResult {
  final bool ok;
  final String message;
  const InviteResult({required this.ok, required this.message});
  factory InviteResult.fromMap(Map<String, dynamic> m) =>
      InviteResult(ok: m['ok'] as bool? ?? false, message: m['message'] as String? ?? '');
}

class SocialService {
  final String _baseUrl;
  SocialService()
      : _baseUrl = const String.fromEnvironment(
          'API_BASE_URL',
          defaultValue: 'https://vitamind-woad.vercel.app',
        );

  Future<Map<String, String>> _authHeaders() async {
    final session = Supabase.instance.client.auth.currentSession;
    if (session == null) throw Exception('Not authenticated');
    return {
      'Authorization': 'Bearer ${session.accessToken}',
      'Content-Type': 'application/json',
    };
  }

  Future<List<SocialConnection>> getConnections() async {
    try {
      final dio = Dio();
      final res = await dio.get(
        '$_baseUrl/api/v1/social/friends',
        options: Options(headers: await _authHeaders()),
      );
      return (res.data['data'] as List? ?? [])
          .map((e) => SocialConnection.fromMap(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      debugPrint('SocialService.getConnections failed: $e');
      rethrow;
    }
  }

  Future<List<FriendActivity>> getFeed() async {
    try {
      final dio = Dio();
      final res = await dio.get(
        '$_baseUrl/api/v1/social/feed',
        options: Options(headers: await _authHeaders()),
      );
      return (res.data['data'] as List? ?? [])
          .map((e) => FriendActivity.fromMap(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      debugPrint('SocialService.getFeed failed: $e');
      rethrow;
    }
  }

  Future<InviteResult> sendInvite(String email) async {
    try {
      final dio = Dio();
      final res = await dio.post(
        '$_baseUrl/api/v1/social/friends',
        data: {'email': email},
        options: Options(headers: await _authHeaders()),
      );
      return InviteResult.fromMap(res.data['data'] as Map<String, dynamic>);
    } catch (e) {
      debugPrint('SocialService.sendInvite failed: $e');
      rethrow;
    }
  }

  Future<void> acceptInvite(String id) async {
    try {
      final dio = Dio();
      await dio.put(
        '$_baseUrl/api/v1/social/friends/$id',
        options: Options(headers: await _authHeaders()),
      );
    } catch (e) {
      debugPrint('SocialService.acceptInvite failed: $e');
      rethrow;
    }
  }

  Future<void> removeConnection(String id) async {
    try {
      final dio = Dio();
      await dio.delete(
        '$_baseUrl/api/v1/social/friends/$id',
        options: Options(headers: await _authHeaders()),
      );
    } catch (e) {
      debugPrint('SocialService.removeConnection failed: $e');
      rethrow;
    }
  }
}

File: apps/web/src/features/social/services/social.service.ts
import { createAdminClient } from '@/lib/supabase/admin'

export interface SocialConnection {
  id: string
  friend_id: string
  friend_name: string
  friend_email: string
  friend_avatar: string | null
  status: 'pending' | 'accepted' | 'blocked'
  direction: 'outgoing' | 'incoming'
  created_at: string
}

export interface FriendActivity {
  friend_id: string
  friend_name: string
  friend_avatar: string | null
  habits_today: number
  tasks_today: number
  momentum_score: number | null
  streak_highlight: string | null   // e.g. "7-day meditation streak"
  updated_at: string
}

export class SocialService {
  // ── Connection management ─────────────────────────────────────────────────

  static async sendInvite(requesterId: string, email: string): Promise<{ ok: boolean; message: string }> {
    const supabase = createAdminClient()

    // Look up user by email
    const { data: target } = await supabase
      .from('users')
      .select('id, name, email')
      .eq('email', email.toLowerCase().trim())
      .single()

    if (!target) {
      return { ok: false, message: 'No VitaMind account found with that email address.' }
    }
    if (target.id === requesterId) {
      return { ok: false, message: 'You cannot add yourself.' }
    }

    // Check if connection already exists (either direction)
    const { data: existing } = await supabase
      .from('social_connections')
      .select('id, status')
      .or(`and(requester_id.eq.${requesterId},addressee_id.eq.${target.id}),and(requester_id.eq.${target.id},addressee_id.eq.${requesterId})`)
      .single()

    if (existing) {
      if (existing.status === 'accepted') return { ok: false, message: 'You are already friends.' }
      if (existing.status === 'pending') return { ok: false, message: 'Invite already pending.' }
    }

    await supabase.from('social_connections').insert({
      requester_id: requesterId,
      addressee_id: target.id,
      status: 'pending',
    })

    return { ok: true, message: `Invite sent to ${target.name}.` }
  }

  static async acceptInvite(userId: string, connectionId: string): Promise<void> {
    const supabase = createAdminClient()
    await supabase
      .from('social_connections')
      .update({ status: 'accepted', updated_at: new Date().toISOString() })
      .eq('id', connectionId)
      .eq('addressee_id', userId)
      .eq('status', 'pending')
  }

  static async removeConnection(userId: string, connectionId: string): Promise<void> {
    const supabase = createAdminClient()
    await supabase
      .from('social_connections')
      .delete()
      .eq('id', connectionId)
      .or(`requester_id.eq.${userId},addressee_id.eq.${userId}`)
  }

  static async getConnections(userId: string): Promise<SocialConnection[]> {
    const supabase = createAdminClient()

    const { data: rows } = await supabase
      .from('social_connections')
      .select('id, requester_id, addressee_id, status, created_at')
      .or(`requester_id.eq.${userId},addressee_id.eq.${userId}`)
      .neq('status', 'blocked')
      .order('created_at', { ascending: false })

    if (!rows?.length) return []

    // Collect friend IDs
    const friendIds = rows.map((r) =>
      r.requester_id === userId ? r.addressee_id : r.requester_id
    )

    const { data: users } = await supabase
      .from('users')
      .select('id, name, email, avatar_url')
      .in('id', friendIds)

    const userMap = new Map((users ?? []).map((u) => [u.id as string, u]))

    return rows.map((r) => {
      const isRequester = r.requester_id === userId
      const friendId = isRequester ? r.addressee_id : r.requester_id
      const friend = userMap.get(friendId)
      return {
        id: r.id as string,
        friend_id: friendId as string,
        friend_name: (friend?.name ?? 'Unknown') as string,
        friend_email: (friend?.email ?? '') as string,
        friend_avatar: (friend?.avatar_url ?? null) as string | null,
        status: r.status as 'pending' | 'accepted' | 'blocked',
        direction: (isRequester ? 'outgoing' : 'incoming') as 'outgoing' | 'incoming',
        created_at: r.created_at as string,
      }
    })
  }

  // ── Activity feed ─────────────────────────────────────────────────────────

  static async getFriendFeed(userId: string): Promise<FriendActivity[]> {
    const supabase = createAdminClient()

    // Get accepted friend IDs
    const { data: conns } = await supabase
      .from('social_connections')
      .select('requester_id, addressee_id')
      .or(`requester_id.eq.${userId},addressee_id.eq.${userId}`)
      .eq('status', 'accepted')

    if (!conns?.length) return []

    const friendIds = conns.map((c) =>
      c.requester_id === userId ? c.addressee_id : c.requester_id
    ) as string[]

    const today = new Date().toISOString().split('T')[0]

    const [{ data: users }, { data: habitLogs }, { data: tasks }, { data: momentumRows }] = await Promise.all([
      supabase.from('users').select('id, name, avatar_url').in('id', friendIds),
      supabase.from('habit_logs').select('user_id, status').in('user_id', friendIds).eq('date', today).eq('status', 'completed'),
      supabase.from('tasks').select('user_id, status').in('user_id', friendIds).eq('status', 'completed').gte('completed_at', `${today}T00:00:00Z`),
      supabase.from('momentum_snapshots').select('user_id, score').in('user_id', friendIds).order('created_at', { ascending: false }).limit(friendIds.length * 2),
    ])

    const userMap = new Map((users ?? []).map((u) => [u.id as string, u]))
    const habitMap = new Map<string, number>()
    const taskMap = new Map<string, number>()
    const momentumMap = new Map<string, number>()

    for (const h of habitLogs ?? []) {
      habitMap.set(h.user_id as string, (habitMap.get(h.user_id as string) ?? 0) + 1)
    }
    for (const t of tasks ?? []) {
      taskMap.set(t.user_id as string, (taskMap.get(t.user_id as string) ?? 0) + 1)
    }
    for (const m of momentumRows ?? []) {
      if (!momentumMap.has(m.user_id as string)) {
        momentumMap.set(m.user_id as string, m.score as number)
      }
    }

    return friendIds.map((fid) => {
      const u = userMap.get(fid)
      const habits = habitMap.get(fid) ?? 0
      const tasks = taskMap.get(fid) ?? 0
      return {
        friend_id: fid,
        friend_name: (u?.name ?? 'Unknown') as string,
        friend_avatar: (u?.avatar_url ?? null) as string | null,
        habits_today: habits,
        tasks_today: tasks,
        momentum_score: momentumMap.get(fid) ?? null,
        streak_highlight: habits > 0 ? `${habits} habit${habits > 1 ? 's' : ''} completed today` : null,
        updated_at: new Date().toISOString(),
      }
    }).sort((a, b) => (b.habits_today + b.tasks_today) - (a.habits_today + a.tasks_today))
  }
}
